File: distribution/python/dqops/client/models/csv_file_format_spec.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.compression_type import CompressionType
from ..models.new_line_character_type import NewLineCharacterType
from ..types import UNSET, Unset

T = TypeVar("T", bound="CsvFileFormatSpec")


@_attrs_define
class CsvFileFormatSpec:
    """
    Attributes:
        all_varchar (Union[Unset, bool]): Option to skip type detection for CSV parsing and assume all columns to be of
            type VARCHAR.
        allow_quoted_nulls (Union[Unset, bool]): Option to allow the conversion of quoted values to NULL values.
        auto_detect (Union[Unset, bool]): Enables auto detection of CSV parameters.
        compression (Union[Unset, CompressionType]):
        no_compression_extension (Union[Unset, bool]): Whether the compression extension is present at the end of the
            file name.
        dateformat (Union[Unset, str]): Specifies the date format to use when parsing dates.
        decimal_separator (Union[Unset, str]): The decimal separator of numbers.
        delim (Union[Unset, str]): Specifies the string that separates columns within each row (line) of the file.
        escape (Union[Unset, str]): Specifies the string that should appear before a data character sequence that
            matches the quote value.
        filename (Union[Unset, bool]): Whether or not an extra filename column should be included in the result.
        header (Union[Unset, bool]): Specifies that the file contains a header line with the names of each column in the
            file.
        hive_partitioning (Union[Unset, bool]): Whether or not to interpret the path as a hive partitioned path.
        ignore_errors (Union[Unset, bool]): Option to ignore any parsing errors encountered - and instead ignore rows
            with errors.
        new_line (Union[Unset, NewLineCharacterType]):
        quote (Union[Unset, str]): Specifies the quoting string to be used when a data value is quoted.
        sample_size (Union[Unset, int]): The number of sample rows for auto detection of parameters.
        skip (Union[Unset, int]): The number of lines at the top of the file to skip.
        timestampformat (Union[Unset, str]): Specifies the date format to use when parsing timestamps.
    """

    all_varchar: Union[Unset, bool] = UNSET
    allow_quoted_nulls: Union[Unset, bool] = UNSET
    auto_detect: Union[Unset, bool] = UNSET
    compression: Union[Unset, CompressionType] = UNSET
    no_compression_extension: Union[Unset, bool] = UNSET
    dateformat: Union[Unset, str] = UNSET
    decimal_separator: Union[Unset, str] = UNSET
    delim: Union[Unset, str] = UNSET
    escape: Union[Unset, str] = UNSET
    filename: Union[Unset, bool] = UNSET
    header: Union[Unset, bool] = UNSET
    hive_partitioning: Union[Unset, bool] = UNSET
    ignore_errors: Union[Unset, bool] = UNSET
    new_line: Union[Unset, NewLineCharacterType] = UNSET
    quote: Union[Unset, str] = UNSET
    sample_size: Union[Unset, int] = UNSET
    skip: Union[Unset, int] = UNSET
    timestampformat: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        all_varchar = self.all_varchar
        allow_quoted_nulls = self.allow_quoted_nulls
        auto_detect = self.auto_detect
        compression: Union[Unset, str] = UNSET
        if not isinstance(self.compression, Unset):
            compression = self.compression.value

        no_compression_extension = self.no_compression_extension
        dateformat = self.dateformat
        decimal_separator = self.decimal_separator
        delim = self.delim
        escape = self.escape
        filename = self.filename
        header = self.header
        hive_partitioning = self.hive_partitioning
        ignore_errors = self.ignore_errors
        new_line: Union[Unset, str] = UNSET
        if not isinstance(self.new_line, Unset):
            new_line = self.new_line.value

        quote = self.quote
        sample_size = self.sample_size
        skip = self.skip
        timestampformat = self.timestampformat

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if all_varchar is not UNSET:
            field_dict["all_varchar"] = all_varchar
        if allow_quoted_nulls is not UNSET:
            field_dict["allow_quoted_nulls"] = allow_quoted_nulls
        if auto_detect is not UNSET:
            field_dict["auto_detect"] = auto_detect
        if compression is not UNSET:
            field_dict["compression"] = compression
        if no_compression_extension is not UNSET:
            field_dict["no_compression_extension"] = no_compression_extension
        if dateformat is not UNSET:
            field_dict["dateformat"] = dateformat
        if decimal_separator is not UNSET:
            field_dict["decimal_separator"] = decimal_separator
        if delim is not UNSET:
            field_dict["delim"] = delim
        if escape is not UNSET:
            field_dict["escape"] = escape
        if filename is not UNSET:
            field_dict["filename"] = filename
        if header is not UNSET:
            field_dict["header"] = header
        if hive_partitioning is not UNSET:
            field_dict["hive_partitioning"] = hive_partitioning
        if ignore_errors is not UNSET:
            field_dict["ignore_errors"] = ignore_errors
        if new_line is not UNSET:
            field_dict["new_line"] = new_line
        if quote is not UNSET:
            field_dict["quote"] = quote
        if sample_size is not UNSET:
            field_dict["sample_size"] = sample_size
        if skip is not UNSET:
            field_dict["skip"] = skip
        if timestampformat is not UNSET:
            field_dict["timestampformat"] = timestampformat

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        all_varchar = d.pop("all_varchar", UNSET)

        allow_quoted_nulls = d.pop("allow_quoted_nulls", UNSET)

        auto_detect = d.pop("auto_detect", UNSET)

        _compression = d.pop("compression", UNSET)
        compression: Union[Unset, CompressionType]
        if isinstance(_compression, Unset):
            compression = UNSET
        else:
            compression = CompressionType(_compression)

        no_compression_extension = d.pop("no_compression_extension", UNSET)

        dateformat = d.pop("dateformat", UNSET)

        decimal_separator = d.pop("decimal_separator", UNSET)

        delim = d.pop("delim", UNSET)

        escape = d.pop("escape", UNSET)

        filename = d.pop("filename", UNSET)

        header = d.pop("header", UNSET)

        hive_partitioning = d.pop("hive_partitioning", UNSET)

        ignore_errors = d.pop("ignore_errors", UNSET)

        _new_line = d.pop("new_line", UNSET)
        new_line: Union[Unset, NewLineCharacterType]
        if isinstance(_new_line, Unset):
            new_line = UNSET
        else:
            new_line = NewLineCharacterType(_new_line)

        quote = d.pop("quote", UNSET)

        sample_size = d.pop("sample_size", UNSET)

        skip = d.pop("skip", UNSET)

        timestampformat = d.pop("timestampformat", UNSET)

        csv_file_format_spec = cls(
            all_varchar=all_varchar,
            allow_quoted_nulls=allow_quoted_nulls,
            auto_detect=auto_detect,
            compression=compression,
            no_compression_extension=no_compression_extension,
            dateformat=dateformat,
            decimal_separator=decimal_separator,
            delim=delim,
            escape=escape,
            filename=filename,
            header=header,
            hive_partitioning=hive_partitioning,
            ignore_errors=ignore_errors,
            new_line=new_line,
            quote=quote,
            sample_size=sample_size,
            skip=skip,
            timestampformat=timestampformat,
        )

        csv_file_format_spec.additional_properties = d
        return csv_file_format_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/metadata/sources/fileformat/csv/CsvFileFormatSpec.java
package com.dqops.metadata.sources.fileformat.csv;

import com.dqops.connectors.duckdb.config.DuckdbCompressionTypeOption;
import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.basespecs.AbstractSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.id.HierarchyNodeResultVisitor;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.metadata.sources.fileformat.CompressionType;
import com.dqops.metadata.sources.fileformat.TableOptionsFormatter;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import lombok.experimental.FieldNameConstants;

import java.util.List;
import java.util.Objects;

/**
 * Csv file format specification for querying data in the csv format files.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
@FieldNameConstants
public class CsvFileFormatSpec extends AbstractSpec {

    private static final ChildHierarchyNodeFieldMapImpl<CsvFileFormatSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractSpec.FIELDS) {
        {
        }
    };

    @JsonPropertyDescription("Option to skip type detection for CSV parsing and assume all columns to be of type VARCHAR.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean allVarchar;

    @JsonPropertyDescription("Option to allow the conversion of quoted values to NULL values.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean allowQuotedNulls;

    @JsonPropertyDescription("Enables auto detection of CSV parameters.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean autoDetect = true;

    @JsonPropertyDescription("The compression type for the file. By default this will be detected automatically from the file extension (e.g., t.csv.gz will use gzip, t.csv will use none). Options are none, gzip, zstd.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private CompressionType compression;

    @JsonPropertyDescription("Whether the compression extension is present at the end of the file name.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean noCompressionExtension;

    @JsonPropertyDescription("Specifies the date format to use when parsing dates.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String dateformat;

    @JsonPropertyDescription("The decimal separator of numbers.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String decimalSeparator;

    @JsonPropertyDescription("Specifies the string that separates columns within each row (line) of the file.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String delim;

    @JsonPropertyDescription("Specifies the string that should appear before a data character sequence that matches the quote value.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String escape;

    @JsonPropertyDescription("Whether or not an extra filename column should be included in the result.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean filename;

    @JsonPropertyDescription("Specifies that the file contains a header line with the names of each column in the file.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean header;

    @JsonPropertyDescription("Whether or not to interpret the path as a hive partitioned path.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean hivePartitioning;

    @JsonPropertyDescription("Option to ignore any parsing errors encountered - and instead ignore rows with errors.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Boolean ignoreErrors;

    @JsonPropertyDescription("Set the new line character(s) in the file. Options are '\\r','\\n', or '\\r\\n'.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private NewLineCharacterType newLine;

    @JsonPropertyDescription("Specifies the quoting string to be used when a data value is quoted.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String quote;

    @JsonPropertyDescription("The number of sample rows for auto detection of parameters.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Long sampleSize;

    @JsonPropertyDescription("The number of lines at the top of the file to skip.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Long skip;

    @JsonPropertyDescription("Specifies the date format to use when parsing timestamps.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private String timestampformat;

    /**
     * Formats the table options to be used in SQL query. The set (non null) options are added only.
     *
     * @param filePathList The names of files with data.
     * @return The formatted source table with the options.
     */
    public String buildSourceTableOptionsString(List<String> filePathList, TableSpec tableSpec) {
        TableOptionsFormatter tableOptionsFormatter = new TableOptionsFormatter("read_csv", filePathList);
        tableOptionsFormatter.formatValueWhenSet(Fields.allVarchar, allVarchar);
        tableOptionsFormatter.formatValueWhenSet(Fields.allowQuotedNulls, allowQuotedNulls);
        tableOptionsFormatter.formatValueWhenSet(Fields.autoDetect, autoDetect);
        tableOptionsFormatter.formatColumns("columns", tableSpec);
        tableOptionsFormatter.formatStringWhenSet(Fields.compression, DuckdbCompressionTypeOption.fromCompressionType(compression));
        tableOptionsFormatter.formatStringWhenSet(Fields.dateformat, dateformat);
        tableOptionsFormatter.formatStringWhenSet(Fields.decimalSeparator, decimalSeparator);
        tableOptionsFormatter.formatStringWhenSet(Fields.delim, delim);
        tableOptionsFormatter.formatStringWhenSet(Fields.escape, escape);
        tableOptionsFormatter.formatValueWhenSet(Fields.filename, filename);
        tableOptionsFormatter.formatValueWhenSet(Fields.header, header);
        tableOptionsFormatter.formatValueWhenSet(Fields.hivePartitioning, hivePartitioning);
        tableOptionsFormatter.formatValueWhenSet(Fields.ignoreErrors, ignoreErrors);
        tableOptionsFormatter.formatStringWhenSet(Fields.newLine, newLine);
        tableOptionsFormatter.formatStringWhenSet(Fields.quote, quote);
        tableOptionsFormatter.formatValueWhenSet(Fields.sampleSize, sampleSize);
        tableOptionsFormatter.formatValueWhenSet(Fields.skip, skip);
        tableOptionsFormatter.formatStringWhenSet(Fields.timestampformat, timestampformat);
        return tableOptionsFormatter.build();
    }

    /**
     * Returns the option that skip type detection for CSV parsing and assume all columns to be of type VARCHAR.
     *
     * @return All varchar option.
     */
    public Boolean getAllVarchar() {
        return allVarchar;
    }

    /**
     * Sets the option that skip type detection for CSV parsing and assume all columns to be of type VARCHAR.
     *
     * @param allVarchar All varchar.
     */
    public void setAllVarchar(Boolean allVarchar) {
        setDirtyIf(!Objects.equals(this.allVarchar, allVarchar));
        this.allVarchar = allVarchar;
    }

    /**
     * Returns the option that allow the conversion of quoted values to NULL values.
     *
     * @return Allow quoted nulls option.
     */
    public Boolean getAllowQuotedNulls() {
        return allowQuotedNulls;
    }

    /**
     * Sets the option that allow the conversion of quoted values to NULL values.
     *
     * @param allowQuotedNulls Allow quoted nulls.
     */
    public void setAllowQuotedNulls(Boolean allowQuotedNulls) {
        setDirtyIf(!Objects.equals(this.allowQuotedNulls, allowQuotedNulls));
        this.allowQuotedNulls = allowQuotedNulls;
    }

    /**
     * Returns the state of the option that enables auto detection of CSV parameters.
     *
     * @return The auto detect option.
     */
    public Boolean getAutoDetect() {
        return autoDetect;
    }

    /**
     * Sets the option that enables auto detection of CSV parameters.
     *
     * @param autoDetect Auto detect option.
     */
    public void setAutoDetect(Boolean autoDetect) {
        setDirtyIf(!Objects.equals(this.autoDetect, autoDetect));
        this.autoDetect = autoDetect;
    }

    /**
     * Returns the compression type for the file. By default this will be detected automatically from the file extension
     *
     * @return Compression type.
     */
    public CompressionType getCompression() {
        return compression;
    }

    /**
     * Sets the compression type for the file. By default this will be detected automatically from the file extension
     *
     * @param compression Compression type.
     */
    public void setCompression(CompressionType compression) {
        setDirtyIf(!Objects.equals(this.compression, compression));
        this.compression = compression;
    }

    /**
     * Returns the accountName
     * @return accountName.
     */
    public Boolean getNoCompressionExtension() {
        return noCompressionExtension;
    }

    /**
     * Sets noCompressionExtension.
     * @param noCompressionExtension noCompressionExtension.
     */
    public void setNoCompressionExtension(Boolean noCompressionExtension) {
        setDirtyIf(!Objects.equals(this.noCompressionExtension, noCompressionExtension));
        this.noCompressionExtension = noCompressionExtension;
    }

    /**
     * Returns the date format to use when parsing dates.
     *
     * @return Date format.
     */
    public String getDateformat() {
        return dateformat;
    }


    /**
     * Sets the date format to use when parsing dates.
     *
     * @param dateformat Date format.
     */
    public void setDateformat(String dateformat) {
        setDirtyIf(!Objects.equals(this.dateformat, dateformat));
        this.dateformat = dateformat;
    }

    /**
     * Returns the decimal separator of numbers.
     *
     * @return Decimal separator.
     */
    public String getDecimalSeparator() {
        return decimalSeparator;
    }

    /**
     * Sets the decimal separator of numbers.
     *
     * @param decimalSeparator Decimal separator.
     */
    public void setDecimalSeparator(String decimalSeparator) {
        setDirtyIf(!Objects.equals(this.decimalSeparator, decimalSeparator));
        this.decimalSeparator = decimalSeparator;
    }

    /**
     * Returns the string that separates columns within each row (line) of the file.
     *
     * @return Delimiter.
     */
    public String getDelim() {
        return delim;
    }

    /**
     * Sets the string that separates columns within each row (line) of the file.
     *
     * @param delim Delimiter.
     */
    public void setDelim(String delim) {
        setDirtyIf(!Objects.equals(this.delim, delim));
        this.delim = delim;
    }

    /**
     * Returns the string that should appear before a data character sequence that matches the quote value.
     *
     * @return Escape char sequence.
     */
    public String getEscape() {
        return escape;
    }

    /**
     * Sets the string that should appear before a data character sequence that matches the quote value.
     *
     * @param escape Escape char sequence.
     */
    public void setEscape(String escape) {
        setDirtyIf(!Objects.equals(this.escape, escape));
        this.escape = escape;
    }

    /**
     * Returns whether or not an extra filename column should be included in the result.
     *
     * @return The filename option state.
     */
    public Boolean getFilename() {
        return filename;
    }

    /**
     * Sets that an extra filename column should be included in the result.
     *
     * @param filename The filename option state.
     */
    public void setFilename(Boolean filename) {
        setDirtyIf(!Objects.equals(this.filename, filename));
        this.filename = filename;
    }

    /**
     * Returns that the file contains a header line with the names of each column is in the file.
     *
     * @return The header option state.
     */
    public Boolean getHeader() {
        return header;
    }

    /**
     * Sets that the file contains a header line with the names of each column is in the file.
     *
     * @param header The header option state.
     */
    public void setHeader(Boolean header) {
        setDirtyIf(!Objects.equals(this.header, header));
        this.header = header;
    }

    /**
     * Returns whether or not to interpret the path as a hive partitioned path.
     *
     * @return The hive partitioning option state.
     */
    public Boolean getHivePartitioning() {
        return hivePartitioning;
    }

    /**
     * Sets to interpret the path as a hive partitioned path.
     *
     * @param hivePartitioning Hive partitioning option state.
     */
    public void setHivePartitioning(Boolean hivePartitioning) {
        setDirtyIf(!Objects.equals(this.hivePartitioning, hivePartitioning));
        this.hivePartitioning = hivePartitioning;
    }

    /**
     * Returns option that ignore any parsing errors encountered - and instead ignore rows with errors.
     *
     * @return Ignore errors option state.
     */
    public Boolean getIgnoreErrors() {
        return ignoreErrors;
    }

    /**
     * Sets option that ignore any parsing errors encountered - and instead ignore rows with errors.
     *
     * @param ignoreErrors Ignore errors option state.
     */
    public void setIgnoreErrors(Boolean ignoreErrors) {
        setDirtyIf(!Objects.equals(this.ignoreErrors, ignoreErrors));
        this.ignoreErrors = ignoreErrors;
    }

    /**
     * Returns the new line character(s) in the file.
     *
     * @return New line value.
     */
    public NewLineCharacterType getNewLine() {
        return newLine;
    }

    /**
     * Set the new line character(s) in the file. Options are '\r','\n', or '\r\n'.
     *
     * @param newLine New line value.
     */
    public void setNewLine(NewLineCharacterType newLine) {
        setDirtyIf(!Objects.equals(this.newLine, newLine));
        this.newLine = newLine;
    }

    /**
     * Returns the quoting string to be used when a data value is quoted.
     *
     * @return Quoting string.
     */
    public String getQuote() {
        return quote;
    }

    /**
     * Sets the quoting string to be used when a data value is quoted.
     *
     * @param quote Quoting string.
     */
    public void setQuote(String quote) {
        setDirtyIf(!Objects.equals(this.quote, quote));
        this.quote = quote;
    }

    /**
     * Returns the number of sample rows for auto detection of parameters.
     *
     * @return Number of rows for sampling.
     */
    public Long getSampleSize() {
        return sampleSize;
    }

    /**
     * Sets the number of sample rows for auto detection of parameters.
     *
     * @param sampleSize Number of rows for sampling.
     */
    public void setSampleSize(Long sampleSize) {
        setDirtyIf(!Objects.equals(this.sampleSize, sampleSize));
        this.sampleSize = sampleSize;
    }

    /**
     * Returns the number of lines at the top of the file to skip.
     *
     * @return Number of lines to skip.
     */
    public Long getSkip() {
        return skip;
    }

    /**
     * Sets the number of lines at the top of the file to skip.
     *
     * @param skip Number of lines to skip.
     */
    public void setSkip(Long skip) {
        setDirtyIf(!Objects.equals(this.skip, skip));
        this.skip = skip;
    }

    /**
     * Returns the date format to use when parsing timestamps.
     *
     * @return Timestamp format.
     */
    public String getTimestampformat() {
        return timestampformat;
    }

    /**
     * Sets the date format to use when parsing timestamps.
     *
     * @param timestampformat Timestamp format.
     */
    public void setTimestampformat(String timestampformat) {
        setDirtyIf(!Objects.equals(this.timestampformat, timestampformat));
        this.timestampformat = timestampformat;
    }

    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    @Override
    public <P, R> R visit(HierarchyNodeResultVisitor<P, R> visitor, P parameter) {
        return visitor.accept(this, parameter);
    }

    /**
     * Creates and returns a deep clone (copy) of this object.
     */
    @Override
    public CsvFileFormatSpec deepClone() {
        return (CsvFileFormatSpec)super.deepClone();
    }

    /**
     * Creates an expanded and trimmed deep copy of the spec.
     * Configurable properties will be expanded if they contain environment variables or secrets.
     *
     * @param secretValueProvider Secret value provider.
     * @param lookupContext       Secret value lookup context used to access shared credentials.
     * @return Cloned, trimmed and expanded table specification.
     */
    public CsvFileFormatSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        CsvFileFormatSpec cloned = this.deepClone();
        cloned.dateformat = secretValueProvider.expandValue(cloned.dateformat, lookupContext);
        cloned.decimalSeparator = secretValueProvider.expandValue(cloned.decimalSeparator, lookupContext);
        cloned.delim = secretValueProvider.expandValue(cloned.delim, lookupContext);
        cloned.escape = secretValueProvider.expandValue(cloned.escape, lookupContext);
        cloned.quote = secretValueProvider.expandValue(cloned.quote, lookupContext);
        cloned.timestampformat = secretValueProvider.expandValue(cloned.timestampformat, lookupContext);
        return cloned;
    }
}
